File: backend/fabric/emissions-chaincode/emissions.go
package main

import (
	"encoding/json"
	"fmt"
	"log"
	"time"

	"github.com/hyperledger/fabric-contract-api-go/contractapi"
)

// SmartContract provides functions for managing emissions
type SmartContract struct {
	contractapi.Contract
}

// EmissionRecord describes the CO2 data
type EmissionRecord struct {
	ID            string  `json:"id"`
	CorrelationID string  `json:"correlationId"`
	PlantID       string  `json:"plantId"`
	SupplierID    string  `json:"supplierId"`
	Stage         string  `json:"stage"`
	CO2Tonnes     float64 `json:"co2Tonnes"`
	Timestamp     string  `json:"timestamp"`
	Verifier      string  `json:"verifier"`
}

// InitLedger adds a base set of emissions to the ledger
func (s *SmartContract) InitLedger(ctx contractapi.TransactionContextInterface) error {
	return nil
}

// RecordEmission adds a new emission record to the world state
func (s *SmartContract) RecordEmission(ctx contractapi.TransactionContextInterface, id string, correlationId string, plantId string, supplierId string, stage string, co2 float64) error {
	exists, err := s.EmissionExists(ctx, id)
	if err != nil {
		return err
	}
	if exists {
		return fmt.Errorf("the emission record %s already exists", id)
	}

	record := EmissionRecord{
		ID:            id,
		CorrelationID: correlationId,
		PlantID:       plantId,
		SupplierID:    supplierId,
		Stage:         stage,
		CO2Tonnes:     co2,
		Timestamp:     time.Now().Format(time.RFC3339),
		Verifier:      "DigitalTwinEngine_Fabric_v1",
	}
	recordJSON, err := json.Marshal(record)
	if err != nil {
		return err
	}

	return ctx.GetStub().PutState(id, recordJSON)
}

// QueryEmission returns the emission record stored in the world state with given id
func (s *SmartContract) QueryEmission(ctx contractapi.TransactionContextInterface, id string) (*EmissionRecord, error) {
	recordJSON, err := ctx.GetStub().GetState(id)
	if err != nil {
		return nil, fmt.Errorf("failed to read from world state: %v", err)
	}
	if recordJSON == nil {
		return nil, fmt.Errorf("the emission record %s does not exist", id)
	}

	var record EmissionRecord
	err = json.Unmarshal(recordJSON, &record)
	if err != nil {
		return nil, err
	}

	return &record, nil
}

// EmissionExists returns true when asset with given ID exists in world state
func (s *SmartContract) EmissionExists(ctx contractapi.TransactionContextInterface, id string) (bool, error) {
	recordJSON, err := ctx.GetStub().GetState(id)
	if err != nil {
		return false, fmt.Errorf("failed to read from world state: %v", err)
	}

	return recordJSON != nil, nil
}

// GetAllEmissions returns all emission records found in world state
func (s *SmartContract) GetAllEmissions(ctx contractapi.TransactionContextInterface) ([]*EmissionRecord, error) {
	resultsIterator, err := ctx.GetStub().GetStateByRange("", "")
	if err != nil {
		return nil, err
	}
	defer resultsIterator.Close()

	var records []*EmissionRecord
	for resultsIterator.HasNext() {
		queryResponse, err := resultsIterator.Next()
		if err != nil {
			return nil, err
		}

		var record EmissionRecord
		err = json.Unmarshal(queryResponse.Value, &record)
		if err != nil {
			return nil, err
		}
		records = append(records, &record)
	}

	return records, nil
}

func main() {
	emissionChaincode, err := contractapi.NewChaincode(&SmartContract{})
	if err != nil {
		log.Panicf("Error creating emissions chaincode: %v", err)
	}

	if err := emissionChaincode.Start(); err != nil {
		log.Panicf("Error starting emissions chaincode: %v", err)
	}
}

File: backend/fabric/emissions-chaincode/lib/emissions.js
'use strict';

const { Contract } = require('fabric-contract-api');

class EmissionsContract extends Contract {

    async initLedger(ctx) {
        // No initial assets needed
    }

    async RecordEmission(ctx, correlationId, plantId, supplierId, stage, thingId, co2, timestamp, worldStateDigest, payload) {
        // Use a composite key or the correlationId + stage as the ID
        const id = `${correlationId}_${stage}_${thingId}`;
        const exists = await this.EmissionExists(ctx, id);
        if (exists) {
            throw new Error(`The emission record ${id} already exists`);
        }

        const record = {
            id: id,
            correlationId: correlationId,
            plantId: plantId,
            supplierId: supplierId,
            stage: stage,
            thingId: thingId,
            co2Tonnes: parseFloat(co2),
            timestamp: timestamp,
            worldStateDigest: worldStateDigest,
            payload: JSON.parse(payload),
            verifier: 'DigitalTwinEngine_Fabric_v1',
        };

        await ctx.stub.putState(id, Buffer.from(JSON.stringify(record)));
    }

    async QueryEmission(ctx, id) {
        const recordJSON = await ctx.stub.getState(id);
        if (!recordJSON || recordJSON.length === 0) {
            throw new Error(`The emission record ${id} does not exist`);
        }
        return recordJSON.toString();
    }

    async EmissionExists(ctx, id) {
        const recordJSON = await ctx.stub.getState(id);
        return recordJSON && recordJSON.length > 0;
    }

    async GetAllEmissions(ctx) {
        const allResults = [];
        const iterator = await ctx.stub.getStateByRange('', '');
        let result = await iterator.next();
        while (!result.done) {
            const strValue = Buffer.from(result.value.value.toString()).toString('utf8');
            let record;
            try {
                record = JSON.parse(strValue);
            } catch (err) {
                console.log(err);
                record = strValue;
            }
            allResults.push(record);
            result = await iterator.next();
        }
        return JSON.stringify(allResults);
    }

    async GetEmissionsByCorrelation(ctx, correlationId) {
        const allResults = [];
        const iterator = await ctx.stub.getStateByRange('', '');
        let result = await iterator.next();
        while (!result.done) {
            const strValue = Buffer.from(result.value.value.toString()).toString('utf8');
            let record;
            try {
                record = JSON.parse(strValue);
                if (record.correlationId === correlationId) {
                    allResults.push(record);
                }
            } catch (err) {
                // Skip if not JSON or doesn't match
            }
            result = await iterator.next();
        }
        return JSON.stringify(allResults);
    }
}

module.exports = EmissionsContract;
